### PR TITLE
Add basic support for readonly properties

### PR DESCRIPTION
**Pull request by genintho**
PHP 8.1 added support for readonly properties. 
https://wiki.php.net/rfc/readonly_properties_v2

```php
class Test {
    public readonly string $prop;
 
    public function __construct(string $prop) {
        $this->prop = $prop;
    }
}
```

```php
class Test {
 
    public function __construct(public readonly string $prop) {
    }
}
```

### CODE REVIEW

**Comment by genintho** [comment]
I would love some feedback to know if this is a good approach.

File: src/parser/class.js [hunk 1]
```diff
@@ -172,19 +172,25 @@ module.exports = {
        */
       function read_variable_declaration() {
         const result = this.node("property");
+        let readonly = false;
+        if (this.text() === "readonly") {
```
Comment (cseufert) [inline src/parser/class.js:176]: You should be able to check for the token T_READ_ONLY here rather than text.

**Review by cseufert — COMMENTED** [review]
You should be able to add the same check just after the promoted properties are found in the parameter parser see: src/parser/function.js:239

Also the other snapshots should be updated since you have updated the Parameter Node

**Comment by genintho** [comment]
@cseufert I followed your suggestion. IMO this is overloading a bit much the concept of "Parameters" and creating a "class constructor" and "class constructor parameter" would be cleaner. But that's way more work :D

**Comment by cseufert** [comment]
@genintho Yes, we have been mostly following how this parser does it: https://github.com/nikic/PHP-Parser

Generally, the nodes and attributes are following the same names. I guess what it really needed is some sort of parse error if we encounter promoted or read-only param and we are not in a constructor, however that's probably outside of the scope of these changes.

**Comment by genintho** [comment]
@cseufert I added something, let me know what you think of the approach

File: src/parser/class.js [hunk 2]
```diff
@@ -122,6 +122,7 @@ module.exports = {
       } else if (
         allow_variables &&
         (this.token === this.tok.T_VARIABLE ||
+          this.token === this.tok.T_READ_ONLY ||
```
Comment (czosel) [inline src/parser/class.js:125]: I think we should add a version check here, such that parsing read only properties only works if `version >= 810`

File: test/snapshot/__snapshots__/graceful.test.js.snap [hunk 3]
```diff
@@ -837,18 +838,19 @@ Program {
           "kind": "propertystatement",
           "properties": Array [
             Property {
-              "attrGroups": Array [],
+              "attrGroups": null,
```
Comment (czosel) [inline test/snapshot/__snapshots__/graceful.test.js.snap:841]: This diff and the one right below looks a bit off, doesn't it?

File: src/parser/function.js [hunk 4]
```diff
@@ -228,15 +230,28 @@ module.exports = {
...
     let parameterName = null;
     let value = null;
     let types = null;
     let nullable = false;
+    let readonly = false;
     let attrs = [];
     if (this.token === this.tok.T_ATTRIBUTE) attrs = this.read_attr_list();
     const flags = this.read_promoted();
+
+    if (this.token === this.tok.T_READ_ONLY) {
```
Comment (czosel) [inline src/parser/function.js:244]: This line needs a version check as well

File: test/snapshot/__snapshots__/graceful.test.js.snap (hunk 3 above)
Comment (genintho) [reply test/snapshot/__snapshots__/graceful.test.js.snap:841]: The test in question is

```javascript
 it("test class", function () {
      expect(test.parseEval("class foo { foo const A = 1 ")).toMatchSnapshot();
    });
```

My understanding of this test suite is to make sure that the parser does not throw an error when providing invalid PHP and when setting the option `suppressErrors` to true. 

If that understanding is correct, I am not sure I need to try to "fix" how an invalid PHP class gets parsed.

Comment (czosel) [reply test/snapshot/__snapshots__/graceful.test.js.snap:841]: Ah, I didn't check what the test was actually doing. Sounds reasonable!

**Review by czosel — APPROVED** [review]
:100:

**Comment by czosel** [comment]
Published in `v3.1.0-beta.2` :tada:

**Comment by genintho** [comment]
Cool Thanks!
Do you know why all the update of this package are published behind a "beta" tag as opposed to a regular version number?

**Comment by czosel** [comment]
Thank you for implementing this! :wink: 

We're currently doing beta releases because quite many changes have landed since the last stable release. If no significant issues are reported in the coming weeks, we'll do a regular release.

**Pull request merged**